### ISSUE TITLE
Write RigidBody / Physics Unit Tests

### ISSUE THREAD

**Issue by ilexp** [AdamsLair/duality#533]
### Summary

There are already unit tests that check some basic cloning / copying functionality, but there are none that test actual physics and transformation behavior. Add a few basic tests to ensure physics works properly.

### Analysis

- The basic test layout could be to create a scene, make it current and simulate a certain number of frames with a fixed timestep, similar to the sandbox single-frame step.
- Tests could include:
  - A ball falling down onto a platform.
  - Non-static bodies not depending on parent transform, unless `AllowParent` was set to true.
  - Probably lots of others, including properties, joints, etc., but keep it short and simple for now.

**Comment by Barsonax** [comment]
I could pick this one up. Already did some unit tests in Pathfindax that used raycasts. I can add the tests to RigidBodyTest right?

Not really sure what you mean with 'Non-static bodies not depending on parent transform, unless AllowParent was set to true.' though

**Comment by ilexp** [comment]
> I could pick this one up. Already did some unit tests in Pathfindax that used raycasts. I can add the tests to RigidBodyTest right?

Yes, that would be a good place. As the number of tests grows, we might think about splitting it up, but not now.

> Not really sure what you mean with 'Non-static bodies not depending on parent transform, unless AllowParent was set to true.' though

If you parent a GameObject that has a RigidBody to another one with a Transform, the child object will ignore changes to its parent Transform to ensure a well-defined physics simulation. This exception will be disabled when switching on `AllowParent`.

Maybe just start with the "A ball falling down onto a platform" one for now, do a PR for that one only, we'll do the reviews and then go from there.

**Comment by ilexp** [comment]
There are existing tests that call `DualityApp.Update`, which do seem to work fine on my machine and the CI server, for example `SceneTest.SwitchToDuringLeave` and others in the same class. Can you step through them and see if you get the same behavior?

**Comment by Barsonax** [comment]
Tested with `SceneTest.SwitchToDuringLeave` and this still happens:
<img width="981" alt="image" src="https://user-images.githubusercontent.com/19387223/31314853-96ec84c0-ac0b-11e7-8b03-0dd8d7a77da4.png">

**Comment by ilexp** [comment]
I'll try to see if I can reproduce this with any of the VS versions I have installed. I'm usually on Visual Studio 2013, maybe this only happens for newer versions.

**Comment by Barsonax** [comment]
Iam using Visual studio 2017 and running the tests using the Resharper test runner.

**Comment by ilexp** [comment]
Debugged the same method in Visual Studio 2017 and stepped over it without issues. Might be an issue related to your dev environment in some way. I don't know the Resharper test runner, but can you check if it works when using Visual Studios test explorer? To get NUnit tests in there, install the NUnit 2 Test Adapter extension via `Tools / Extensions and Updates...`.

**Comment by Barsonax** [comment]
Ah that also explains why I had to use Resharper to actually run the tests. In pathfindax I use the NUnit3TestAdapter nuget package so that you don't need the extension.

Tested it with the Nunit2 extension and now it works without problems. Still maybe consider adding that nuget package to the test projects? That way ppl won't have to install the extension.

**Comment by ilexp** [comment]
> Tested it with the Nunit2 extension and now it works without problems. Still maybe consider adding that nuget package to the test projects? That way ppl won't have to install the extension.

Oh, I wasn't aware that existed. Will take a look!

**Comment by ilexp** [comment]
First test implemented by @Barsonax in the recently linked PR.

**Comment by ilexp** [comment]
### Immediate ToDo

- Implement a unit test that checks PickShapes for correct functionality in both point-based and rect-based overloads, as well as both local and global overloads.
  - Make sure the failure case from issue #291 is included and reproducable with the pre-#579 code.
